tests/draft2019-09/ref.json

```diff
@@ -430,5 +430,109 @@
                 "valid": true
             }
         ]
+    },
+    {
+        "description": "refs with relative uris and defs",
+        "schema": {
+            "$id": "schema1.json",
```

Comment:
I don't agree with that - a schema document can have a relative URI in its top-most $id.

Reply:
https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.2.1
> If present, the value for this keyword MUST be a string, and MUST represent a valid URI-reference. This URI-reference SHOULD be normalized, and _MUST resolve to an absolute-URI_ (without a fragment).

Reply:
`schema1.json` fits that criteria.

If schemeless or hostless $ids were prohibited, lots of other tests in the test suite would be incorrect as well.

Reply:
With another view:

> ### 9.1.1. Initial Base URI
>
> Informatively, the initial base URI of a schema is the URI at which it was found, whether that was a network location, a local filesystem, or any other situation identifiable by a URI of any known scheme.

> ### 8.1.2 The "$id" Keyword
>
> Furthermore, in accordance with RFC 3986 section 5.1.2 regarding encapsulating entities, if an "$id" in a subschema is a relative URI-reference, the base URI for resolving that reference is the URI of the parent schema resource.
>
> If no parent schema object explicitly identifies itself as a resource with "$id", the base URI is that of the entire document, as established by the steps given in the previous section.

It does say so explicitly for subschemas, but I would interpret "resolve to an absolute-URI" as allowing a fall-back of a relative URI to using the "URI at which it was found" also for top-level ones.

I don't know if this is the best place to discuss this, but this is an important use-case for schemas which are hosted locally, therefore their absolute URI is not statically known, but they still refer to each other. It is useful for them to carry their relative URI, which can then be resolved relative to the base URI configured at the client.

Reply:
@karenetheridge How does `schema1.json` fit the definition of absolute-URI? https://tools.ietf.org/html/rfc3986#section-4.3
> absolute-URI  = scheme ":" hier-part [ "?" query ]

@amosonn It's not a fallback. They all apply. URLs resolve in a chain from the topmost context down. Retrieval URL > `$id` > embedded `$id` > `$ref`. If there is a retrieval URL, then a relative URI is ok for `$id` because it will resolve to an absolute URI. In contexts where there isn't a retrieval URL, an `$id` with a relative URI won't resolve to an absolute URI. Technically, it isn't against the spec to not have a retrieval URL or an `$id`, but that would mean that any `$ref`s that reference an external schema could not be relative because there isn't any URI to resolve it against.

Referencing local schemas is well supported. You can use retrieval URLs (`file://`) or use `$id` to identify your schemas. Remember that `$id`s don't need to be network accessible. They are just identifiers. You can make them up. Just put a scheme and domain in front of them,  `http://myapp.com/schema1`.

Reply:
I don't think it's invalid for them to be schemeless. There are plenty of usecases where a schema is evaluated and there is no base URI provided at all (i.e. the base uri is '').

Reply:
@karenetheridge This discussion is going nowhere. Can we at least agree that changing this test to an absolute top level `$id` doesn't affect what is being covered by the test and that an absolute URI is a valid value for `$id`? That way there is nothing controversial and we can allow this PR to move forward. We can argue about what the spec says (and possibly what it should say) elsewhere.

Reply:
sure. I did not mean to get in the middle of this particular PR.

Reply:
I'm actually not quite sure that we can agree on this. Handling non-absolute URI-s is different than handling absolute ones, otherwise whether the spec allows this or not wouldn't be an issue. Part of what I wanted to test here is handling non-absolute ones, in combination with fragment-only ones (i.e. `$refs`), in particular. But if this is deemed too controversial, I can drop those for now. I will check that the original bugs are still found.

Reply:
Ok, will update all top-level `$id`-s to be absolute.